typeahead: Add type definitions for `shared/js/typeahead.js`

PIG208 opened the pull request:
We add a `.d.ts` file for `typeahead.js` so that `people.js` has the types when importing it.
Because Zulip Mobile also relies on the shared module, we do not directly convert `typeahead.js` to TypeScript to avoid transpilation.

This change also comes with some quick refactorings:
- Removed the default value to the `get_item` parameter of `triage`, since that default is incompatible with the type signature we expect.
- Removed `is_unicode_emoji`. So we don't need a user-defined type guard to type narrow `Emoji`, which might have undefined `emoji_code`.

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [ ] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [ ] Each commit is a coherent idea.
- [ ] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

timabbott commented:
@N-Shar-ma @aero31aero can you help review this?

timabbott commented:
Also @gnprice @chrisbobbe thoughts on the question of whether we should avoid changing files under `shared/` to be fully implemented in TypeScript to avoid transpilation? I'm not sure how to think about that potential constraint.

timabbott commented:
@PIG208 this one may take a while to answer the above questions, so please be encouraged to work on other conversions in parallel.

andersk commented:
> Also @gnprice @chrisbobbe thoughts on the question of whether we should avoid changing files under `shared/` to be fully implemented in TypeScript to avoid transpilation? I'm not sure how to think about that potential constraint.

The path forward on that is #22586.

PIG208 commented on static/shared/js/typeahead.js:
```diff
@@ -72,28 +72,28 @@ export function clean_query_lowercase(query) {
     return function (emoji) {
-        const matches_emoji_literal =
-            is_unicode_emoji(emoji) && parse_unicode_emoji_code(emoji.emoji_code) === query;
+        const matches_emoji_literal = maybe_parse_unicode_emoji_code(emoji.emoji_code) === query;
```
That's right, thanks.

timabbott commented on static/shared/js/typeahead.d.ts:
```diff
@@ -0,0 +1,29 @@
+type Emoji = {
+    emoji_name: string;
+    emoji_code?: string;
```
Under what circumstance would `emoji_code` be undefined? I think that all emoji types should have that field set unconditionally.

PIG208 replied:
`emoji_code` is only available for unicode emojis. We don't have that for realm emojis.

timabbott commented:
Posted one question, otherwise lgtm.

PIG208 commented:
I rebased the PR to incorporate changes related to `prioritise_realm_emojis`. I have also reworked the type annotations so that we translate the original `.js` file into a typescript given that #22586 has been merged.

The test cases are updated to always ensure the presence of `reaction_type` on the emoji object to be passed into any of `typeahead`'s helpers. This assumption holds because `emoji_collection` is the only source of emoji objects where we call the emoji matcher/sorter from.

PIG208 commented:
Rebased to resolve the merge conflicts with 431cd8b10a1a020994433fdeca0ea500c183d153.

andersk commented:
We don’t need to change the `parse_unicode_emoji_code` API. We can type it `parse_unicode_emoji_code(emoji: UnicodeEmoji): string` (with the `UnicodeEmoji` type factored out of `Emoji`), and have callers narrow the argument before the call with a `.reaction_type === "unicode_emoji"` check. This seems simpler than checking for `undefined` after the call.

PIG208 commented:
Rebased to revert back to the original `parse_unicode_emoji_code` API that takes a `string`, and passes `emoji.emoi_code` once we type narrow `Emoji` to `UnicodeEmoji`. Factoring out `UnicodeEmoji` now is not necessary, but it's better for readability.

I also updated codepaths that have the `is_realm_emoji` property missing on the emoji object. However, I am not sure if `is_realm_emoji` is necessary, since it is basically `reaction_type == "realm_emoji" || reaction_type == "zulip_extra_emoji"`

timabbott commented:
Merged, thanks @PIG208!

gnprice commented:
> However, I am not sure if `is_realm_emoji` is necessary, since it is basically `reaction_type == "realm_emoji" || reaction_type == "zulip_extra_emoji"`

Yes, well spotted. It looks like that `is_realm_emoji` property was added very recently. It's unnecessary, and worse than that, it is wrong: the name says it should be true only for realm emoji, but in fact it is also true for the Zulip extra emoji.

Because it changes the API of the shared typeahead code, it looks like it probably will also break updating that code in mobile, though I haven't tested that.

It'd be good to take it out and replace it with looking at `reaction_type`.
